factory-reset/eos-factory-reset

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/python3
+import argparse
+import logging
+import subprocess
+import sys
+
+from systemd import journal
+
+# Options list
+# Each stage has the form (short name, long name, description)
+OPTS = [
+    ('D', 'disable', 'fully remove a scheduled factory reset'),
+]
+
+# Stages list
+# Each stage has the form (short name, long name, description), which will be
+# used as command-line arguments for this program. Make sure to check both this
+# list and OPTS when adding a new stage so short or long names don't clash.
+STAGES = [
+    ('u', 'users', 'remove all user accounts (and home directories)'),
+]
+
+
+def yes_or_no(msg):
+    answer = ''
+    while answer not in ['y', 'n']:
+        answer = input(msg + ' ').lower()
+    return answer == 'y'
+
+
+if __name__ == '__main__':
+    desc = 'reboot the machine and perform a factory reset.'
+
+    log = logging.getLogger(__name__)
+    log.propagate = False
+    log.setLevel(logging.INFO)
+    log.addHandler(logging.StreamHandler())
+    log.addHandler(journal.JournalHandler())
+
+    parser = argparse.ArgumentParser(description=desc,
+                                     argument_default=argparse.SUPPRESS)
+    for n, name, desc in OPTS:
+        parser.add_argument('-'+n, '--'+name, help=desc, action='store_true')
+
+    group = parser.add_argument_group('stages to be performed on the next boot',
+                                      'if no stage is selected all stages will'
+                                      ' be scheduled')
+    for n, name, desc in STAGES:
+        group.add_argument('-'+n, '--'+name, help=desc, action='store_true')
+
+    args = vars(parser.parse_args())
+    disable = args.pop('disable', False)
+    if disable:
+        stages = list()
+    else:
+        stages = list(args.keys())
+
+    if len(stages) == 0:
+        for _, name, _ in STAGES:
+            stages.append(name)
+
+    action = ''
+    if not disable:
+        print('Selected stages: %s\n' % ' '.join(stages))
+        print('WARNING: You are about to schedule IRREVERSIBLE changes to be')
+        print('performed on the next boot. If you do not want to schedule these')
+        print('changes for the next boot say \'n\' now.\n')
+        action = 'enable'
+    else:
+        print('You are about to remove the scheduled factory reset process.')
+        print('Unless scheduled again, a factory reset will NOT be performed.\n')
+        action = 'disable'
+    if not yes_or_no('Do you want to proceed [y/n]?'):
+        logging.shutdown()
+        sys.exit(0)
+    print()
+
+    log.info('Requesting to %s stages: %s' % (action, ' '.join(stages)))
+    for stage in stages:
+        unit = 'eos-factory-reset-' + stage + '.service'
+        cmd = ['systemctl', '--quiet', action]
+
+        log.info('Trying to %s %s' % (action, unit))
+        cmd.append(unit)
+        systemctl = subprocess.run(cmd, stderr=subprocess.PIPE)
```

Comment:
If you create a function `def toggle_stage(stage, action)` it would simplify a bit of this and remove duplicated code in the disable-on-failure loop.

Reply:
Agreed.